tests/integration/telemetry/metadataexchange/metadata_exchange_test.go

```diff
@@ -0,0 +1,407 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadataexchange
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	kubeApiCore "k8s.io/api/core/v1"
+
+	"istio.io/istio/pkg/config/protocol"
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/echo/deployment"
+	"istio.io/istio/pkg/test/framework/components/istio"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+	"istio.io/istio/pkg/test/framework/label"
+	"istio.io/istio/pkg/test/framework/resource"
+	"istio.io/istio/pkg/test/kube"
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/istio/tests/integration/telemetry"
+)
+
+var (
+	ist         istio.Instance
+	echoNsInst  namespace.Instance
+	srv         echo.Instance
+	clt         echo.Instance
+	sinkPodName string
+)
+
+type Metadata struct {
+	LogType        string
+	UpstreamPeer   string
+	DownstreamPeer string
+}
+
+func TestHTTPMetadataExchangeBetweenSidecars(t *testing.T) {
+	framework.NewTest(t).
+		Features("observability.telemetry.tracing.server").
+		Run(func(t framework.TestContext) {
+			resetAlsSinkServer(t)
+			err := retry.UntilSuccess(func() error {
+				if err := sendTraffic(clt, http.Header{}, false); err != nil {
+					return err
+				}
+
+				if err := validateMetadataExchangeAmongSidecars(t, "httpMx"); err != nil {
+					return err
+				}
+				t.Logf("http metadata-exchange validated")
+
+				return nil
+			}, retry.Delay(framework.TelemetryRetryDelay), retry.Timeout(framework.TelemetryRetryTimeout))
+			if err != nil {
+				t.Fatalf("test failed: %v", err)
+			}
+		})
+}
+
+func TestTCPMetadataExchangeBetweenSidecars(t *testing.T) {
+	framework.NewTest(t).
+		Features("observability.telemetry.tracing.server").
+		Run(func(t framework.TestContext) {
+			resetAlsSinkServer(t)
+			err := retry.UntilSuccess(func() error {
+				if err := sendTraffic(clt, http.Header{}, true); err != nil {
+					return err
+				}
+
+				if err := validateMetadataExchangeAmongSidecars(t, "tcpMx"); err != nil {
+					return err
+				}
+				t.Logf("tcp metadata-exchange validated")
+
+				return nil
+			}, retry.Delay(framework.TelemetryRetryDelay), retry.Timeout(framework.TelemetryRetryTimeout))
+			if err != nil {
+				t.Fatalf("test failed: %v", err)
+			}
+		})
+}
+
+func TestTCPMetadataExchangeBetweenSidecarAndGateway(t *testing.T) {
```

Comment:
If I run this test with latest `master` istiod, client gets stuck. Following I see in sidecar envoy logs:
```
2022-05-14T07:51:08.678952Z	debug	envoy filter	[C2410] Creating connection to cluster outbound|443||istio-egressgateway.istio-system.svc.cluster.local
2022-05-14T07:51:08.678972Z	debug	envoy misc	Allocating TCP conn pool
2022-05-14T07:51:08.678987Z	debug	envoy pool	trying to create new connection
2022-05-14T07:51:08.678992Z	debug	envoy pool	creating a new connection (connecting=0)
2022-05-14T07:51:08.679027Z	debug	envoy connection	[C2411] connecting to 10.244.0.149:8443
2022-05-14T07:51:08.679112Z	debug	envoy connection	[C2411] connection in progress
2022-05-14T07:51:08.679124Z	debug	envoy conn_handler	[C2410] new connection from 10.244.0.152:46940
2022-05-14T07:51:08.679135Z	debug	envoy connection	[C2411] connected
2022-05-14T07:51:08.679925Z	debug	envoy pool	[C2411] attaching to next stream
2022-05-14T07:51:08.679932Z	debug	envoy pool	[C2411] creating stream
2022-05-14T07:51:08.679942Z	debug	envoy filter	[C2410] TCP:onUpstreamEvent(), requestedServerName: edition.cnn.com
2022-05-14T07:51:08.703348Z	debug	envoy main	flushing stats
2022-05-14T07:51:08.714870Z	debug	envoy filter	Alpn Protocol matched. Waiting to read more initial header.
```
looks like client envoy keeps waiting and connection does not proceed because at gateway tcp-mx filter is not there to send the magic number in the application data

Then I reran the test with istiod(`c9c09d8620`), which is before adding `istio-peer-exchange` alpn at gateway, this test and http test sidecar to gateway works fine.